ProcessLib/TwoPhaseFlowWithPrho/TwoPhaseFlowWithPrhoLocalAssembler-impl.h

```diff
@@ -0,0 +1,232 @@
+
+    SpatialPosition pos;
+    pos.setElementID(_element.getID());
+    _process_data._material->setMaterialID(pos);
+
+    const Eigen::MatrixXd& perm = _process_data._material->getPermeability(
+        t, pos, _element.getDimension());
+    assert(perm.rows() == GlobalDim || perm.rows() == 1);
+    GlobalDimMatrixType permeability =
+        GlobalDimMatrixType::Zero(GlobalDim, GlobalDim);
+    if (perm.rows() == GlobalDim)
+        permeability = perm;
+    else if (perm.rows() == 1)
+        permeability.diagonal().setConstant(perm(0, 0));
+    // Note: currently only isothermal case is considered, so the temperature is
+    // assumed to be const
+    // the variation of temperature will be taken into account in future
+    for (unsigned ip = 0; ip < n_integration_points; ip++)
+    {
+        auto const& sm = _shape_matrices[ip];
+
+        double pl_int_pt = 0.;
+        double totalrho_int_pt =
+            0.;  // total mass density of the light component
+        NumLib::shapeFunctionInterpolate(local_x, sm.N, pl_int_pt,
+                                         totalrho_int_pt);
+        auto const& wp = _integration_method.getWeightedPoint(ip);
+        const double integration_factor =
+            sm.integralMeasure * sm.detJ * wp.getWeight();
+
+        double const rho_gas =
+            _process_data._material->getGasDensity(pl_int_pt, _temperature);
+        double const rho_h2o =
+            _process_data._material->getLiquidDensity(pl_int_pt, _temperature);
+
+        double& Sw = _ip_data[ip]._sw;
+        double const X_h2_nonwet = 1.0;            // TODO
+        double& rho_h2_wet = _ip_data[ip]._rho_m;
+        double& dSwdP_gp = _ip_data[ip]._dsw_dpg;
+        double& dSwdrho_gp = _ip_data[ip]._dsw_drho;
+        double& drhoh2wet_dp = _ip_data[ip]._drhom_dpg;
+        double& drhoh2wet_drho = _ip_data[ip]._drhom_drho;
+        if (!_ip_data[ip]._mat_property.computeConstitutiveRelation(
+                t,
+                pos,
+                pl_int_pt,
+                totalrho_int_pt,
+                Sw,
+                rho_h2_wet,
+                dSwdP_gp,
+                dSwdrho_gp,
+                drhoh2wet_dp,
+                drhoh2wet_drho))
+            OGS_FATAL("Computation of local constitutive relation failed.");
+        double const pc = _process_data._material->getCapillaryPressure(
+            t, pos, pl_int_pt, _temperature, Sw);
+
+        double const rho_wet = rho_h2o + rho_h2_wet;
+        _saturation[ip] = Sw;
+        _pressure_nonwetting[ip] = pl_int_pt + pc;
+
+        // Assemble M matrix
+        // nonwetting
+        double dPC_dSw_gp = _process_data._material->getDerivCapillaryPressure(
+            t, pos, pl_int_pt, _temperature, Sw);
+
+        double const porosity = _process_data._material->getPorosity(
+            t, pos, pl_int_pt, _temperature, 0);
+        mass_operator.noalias() = sm.N.transpose() * sm.N * integration_factor;
```

Comment:
(cont'd from integration point data comment)
... especialy those which are expensive to compute...